ref/src/yppm_driver.f90

```diff
@@ -327,4 +331,234 @@ subroutine write_state(filename)
 
   end subroutine write_state
 
+  !------------------------------------------------------------------
+  ! interpolate_state
+  !
+  ! Increase the database by interpFactor.
+  !------------------------------------------------------------------
+  subroutine interpolate_state(interpFactor)
+ 
+  integer, intent(in) :: interpFactor
+
+  ! locals
+  ! Define new variables to replace the old variables.
+  ! Once the new variables are defined, they will replace the 
+  ! old variables. 
+
+  integer new_isd 
+  integer new_ied 
+  integer new_jsd 
+  integer new_jed 
+  integer new_js 
+  integer new_je 
+  real, allocatable :: new_cry(:,:)
+  real, allocatable :: new_q(:,:)
+  real, allocatable :: new_fy2(:,:)
+  real, allocatable :: new_dya(:,:)
+  integer :: odims(4),idims(4)
+
+  ! We'll assume that the interpolation factor is greater or equal to one.
+  ! An interpolation factor of 1 means 1 new interpolated element.
+  ! So for example, a 3x3 matrix with interpFactor=1 becomes a 5x5 matrix
+  ! [ x x x ]     ==> [x o x o x]
+  ! [ x x x ]         [o o o o o]
+  ! [ x x x ]         [x o x o x]
+  !                   [o o o o o]
+  !                   [x o x o x]
+  !
+  ! And an interpFactor=2, means a 3x3 matrix becomes a 7x7 matrix.
+  ! [ x x x ]     ==> [x o o x o o x]
+  ! [ x x x ]         [o o o o o o o]
+  ! [ x x x ]         [o o o o o o o]
+  !                   [x o o x o o o]
+  !                   [o o o o o o o]
+  !                   [o o o o o o o]
+  !                   [x o o x o o x]
+  ! We'll be interpolating on the horizonal, then inperpolating on 
+  ! the vertical.  That way, all the "o"s are interpolated.
```

Comment:
@christopherwharrop-noaa  This is a linear interpolation, not a bi-linear interpolation.  A linear interpolation just needs two known points to interpolate an in-between point.  A bi-linear interpolation needs four known points to interpolate an in-between point.  See https://en.wikipedia.org/wiki/Bilinear_interpolation  I choose the simplest interpolation.  We could do a bi-linear interpolation.

Yes, the code assumes an evenly spaced grid.

Reply:
I understand the bilinear interpolation (I implemented in for the Shallow Water JEDI inerface).  What I don't understand is how, given this,

```
x o o o x
o o o o o
o o o o o
o o o o o
x o o o x
```

the interior o's are interpolated with only two known points (which two points do you choose for those interior points?) I don't see how interpolation in 2D can be done with only two points.  A bi-linear method is what you need I think.  And your code says interpolate one way, and then the other, which is exactly what a bi-linear interpolation is supposed to do, but I can't follow it.  So I am confused.

Reply:
Given your example:

x o o o x
o o o o o
o o o o o
o o o o o
x o o o x

The horizontal pass does this:

x  x  x  x x
o o o o o
o o o o o
o o o o o
x  x  x  x x

And the vertical pass does this:

x  x  x  x x
x  x  x  x x
x  x  x  x x
x  x  x  x x
x  x  x  x x

So that I just need two known and or newly interpolated points.  I can revise to a bi-linear interpolation.  In fact, since I separated the interpolation into its own module, we can apply
any interpolation.

Reply:
Ok.  I see.  I'd very much prefer a bi-linear interpolation because it should be a lot more accurate, calculating values in 2D based on values of surrounding points weighted by distance.

Reply:
Done.